lib/reek/smells/smell_repository.rb

```diff
@@ -42,12 +50,20 @@ def examine(scope)
 
       private
 
+      attr_writer :detectors, :typed_detectors, :configuration, :source_via, :smell_types
+
+      def initialize_detectors
+        smell_types.each do |klass|
+          detectors[klass] = klass.new(source_via)
+        end
+      end
+
       def smell_listeners
-        unless @typed_detectors
-          @typed_detectors = Hash.new { |hash, key| hash[key] = [] }
-          @detectors.each_value { |detector| detector.register(@typed_detectors) }
+        unless typed_detectors
```

Comment:
This was basically a nice memoization pattern, but adding writers makes it more complicated.